### PR TITLE
added models: ShoppingCart and CartItem; added corresponding liquibas…

### PR DESCRIPTION
**Pull request by YuraKhilko**
…e changes; added ShoppingCartRepository and CartItemRepository; added ShoppingCartService with its implementation; added ShoppingCart creation while User registration; added ShoppingCartController with necessary DTO and mappers.

### CODE REVIEW

File: src/main/java/bookstore/model/ShoppingCart.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
+@Where(clause = "is_deleted = FALSE")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    //@GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (sokimaaa) [inline src/main/java/bookstore/model/ShoppingCart.java]: Redundant comment in ShoppingCart.class.

File: src/main/java/bookstore/model/ShoppingCart.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    private Long id;
+    @OneToOne(cascade = CascadeType.ALL)
+    @MapsId
+    @PrimaryKeyJoinColumn(name = "user_id")
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
+    private Set<CartItem> cartItems = new HashSet<>();
+    @Column(name = "is_deleted", nullable = false)
+    private Boolean isDeleted = false;
```
Comment (sokimaaa) [inline src/main/java/bookstore/model/ShoppingCart.java]: I think the `isDeleted` field is redundant here.
 
For example, if the user has been deleted, so our shopping cart has a reference to a deleted user that becomes inactive (so, no need to soft delete the shopping cart)

File: src/main/java/bookstore/repository/ShoppingCartRepository.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package bookstore.repository;
+
+import bookstore.model.ShoppingCart;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    //@EntityGraph(attributePaths = "cartItems")
```
Comment (sokimaaa) [inline src/main/java/bookstore/repository/ShoppingCartRepository.java]: redundant comment

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    public ShoppingCartResponseDto addBookToShoppingCart(
+            Authentication authentication,
+            @RequestBody @Valid CreateCartItemRequestDto requestDto
+    ) {
+        String email = authentication.getName();
+        return shoppingCartService.addCartItem(email, requestDto);
+    }
+
+    @Operation(summary = "Update item of user's shopping cart",
+            description = "Update item of user's shopping cart")
```
Comment (sokimaaa) [inline src/main/java/bookstore/controller/ShoppingCartController.java]: not a clear description of the endpoint. 
what exactly do you update? which data do you need for?

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            @PathVariable Long cartItemId,
+            @RequestBody @Valid UpdateCartItemRequestDto requestDto
+    ) {
+        String email = authentication.getName();
+        return shoppingCartService.updateCartItem(email, cartItemId, requestDto);
+    }
+
+    @Operation(summary = "Delete item from user's shopping cart",
+            description = "Delete item from user's shopping cart")
+    @DeleteMapping("cart-items/{cartItemId}")
```
Comment (sokimaaa) [inline src/main/java/bookstore/controller/ShoppingCartController.java:67]: Bad naming of PUT and DELETE endpoint. 
Was: "cart/cart-items/{cartItemId}"
Suggested: "cart/items/{itemId}"
Why? We already have information that we work with `cart`, so no need to duplicate this info in uri.

Comment (YuraKhilko) [reply src/main/java/bookstore/controller/ShoppingCartController.java:67]: Endpoints' naming have been fixed. 
I agree with you about reudndant dublicate "cart" word in endpoints, but we have such requirements in our task.

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    @Operation(summary = "Add book to user's shopping cart",
+            description = "Get user's shopping cart with cart items")
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto addBookToShoppingCart(
+            Authentication authentication,
+            @RequestBody @Valid CreateCartItemRequestDto requestDto
+    ) {
+        String email = authentication.getName();
+        return shoppingCartService.addCartItem(email, requestDto);
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/ShoppingCartController.java]: **For all methods** - lets use id, not email
```suggestion
        User user = (User) authentication.getPrincipal();
        return shoppingCartService.getById(user.getId());
```

File: src/main/java/bookstore/dto/cartitem/CartItemResponseDtoForShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package bookstore.dto.cartitem;
+
+import lombok.Data;
+
+@Data
+public class CartItemResponseDtoForShoppingCart {
```
Comment (Rommelua) [inline src/main/java/bookstore/dto/cartitem/CartItemResponseDtoForShoppingCart.java]:
```suggestion
public class CartItemResponseDto {
```

File: src/main/java/bookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CreateCartItemRequestDto {
+    @Min(1)
+    private Long bookId;
+    @Min(1)
+    private int quantity;
```
Comment (Rommelua) [inline src/main/java/bookstore/dto/cartitem/CreateCartItemRequestDto.java]: add @NotNull

File: src/main/java/bookstore/dto/cartitem/UpdateCartItemRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package bookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemRequestDto {
+    @Min(1)
+    private int quantity;
```
Comment (Rommelua) [inline src/main/java/bookstore/dto/cartitem/UpdateCartItemRequestDto.java:11]: add @NotNull

File: src/main/java/bookstore/mapper/CartItemMapper.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+    @Mapping(target = "book", ignore = true)
+    CartItem toEntity(CreateCartItemRequestDto requestDto);
+
+    @Mapping(target = "book", ignore = true)
+    CartItem toEntity(UpdateCartItemRequestDto requestDto);
+
+    @AfterMapping
+    default void setBookId(@MappingTarget CartItem cartItem, CreateCartItemRequestDto requestDto) {
```
Comment (Rommelua) [inline src/main/java/bookstore/mapper/CartItemMapper.java]:
```suggestion
    default void setBook(@MappingTarget CartItem cartItem, CreateCartItemRequestDto requestDto) {
```

File: src/main/java/bookstore/mapper/ShoppingCartMapper.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import bookstore.dto.shoppingcart.ShoppingCartResponseDto;
+import bookstore.model.CartItem;
+import bookstore.model.ShoppingCart;
+import java.util.HashSet;
+import java.util.Set;
+import org.mapstruct.AfterMapping;
+import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
+
+@Mapper(config = MapperConfig.class)
```
Comment (Rommelua) [inline src/main/java/bookstore/mapper/ShoppingCartMapper.java]:
```suggestion
@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
```

File: src/main/java/bookstore/mapper/ShoppingCartMapper.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+public interface ShoppingCartMapper {
+    ShoppingCartResponseDto toDto(ShoppingCart shoppingCart);
```
Comment (Rommelua) [inline src/main/java/bookstore/mapper/ShoppingCartMapper.java:13]:
```suggestion
    @Mapping(source = "user.id", target = "userId")
    @Mapping(source = "cartItems", target = "cartItems")
    ShoppingCartResponseDto toDto(ShoppingCart shoppingCart);
```

File: src/main/java/bookstore/mapper/ShoppingCartMapper.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        for (CartItem item: shoppingCart.getCartItems()) {
+            CartItemResponseDtoForShoppingCart cartItem = new CartItemResponseDtoForShoppingCart();
+            cartItem.setId(item.getId());
+            cartItem.setBookId(item.getBook().getId());
+            cartItem.setBookTitle(item.getBook().getTitle());
+            cartItem.setQuantity(item.getQuantity());
+            cartItemSet.add(cartItem);
+        }
+        shoppingCartResponseDto.setCartItems(cartItemSet);
+    }
```
Comment (Rommelua) [inline src/main/java/bookstore/mapper/ShoppingCartMapper.java]:
```suggestion

```

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by Rommelua — APPROVED** [review]

**Pull request merged**